metrics/metrics.go

```diff
@@ -115,12 +201,18 @@ func (cmm *csiMetricsManager) GetRegistry() metrics.KubeRegistry {
 // operationName - Name of the CSI operation.
 // operationErr - Error, if any, that resulted from execution of operation.
 // operationDuration - time it took for the operation to complete
+// labelValues - for each additional label that was defined in WithLabels(), one value has to be passed (usually none)
 func (cmm *csiMetricsManager) RecordMetrics(
 	operationName string,
 	operationErr error,
-	operationDuration time.Duration) {
-	cmm.csiOperationsLatencyMetric.WithLabelValues(
-		cmm.driverName, operationName, getErrorCode(operationErr)).Observe(operationDuration.Seconds())
+	operationDuration time.Duration,
+	labelValues ...string) {
```

Comment:
> So the order of the labels has to be in the same order as the keys?

Yes.

> Would it be better to have a map passed in and order by keys? Should we validate that the length of the values is the same as the length of the keys?

We could make the call more robust against incorrect usage; both of this would work. Map corresponds to the `GetMetricWith` method (https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec.GetMetricWith), checking the length to `With` (https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec.With). Either way, we then also should introduce an error return value and expect that the callers handle this.

I don't have a strong preference either way.

Reply:
One further comment: `labelValues ...string` was also chosen because it's compatible with current code which doesn't pass any additional labels. We could introduce `WithLabels` and/or `WithLabelValues` methods that return a CSIMetricsManager which uses the additional labels when its RecordMetrics is invoked. RecordMetrics itself then has the same parameters as it has before this PR.

This is probably the best solution, but also the most complicated one. I'm wondering whether it is overkill.

@msau42 do you have any preferences?

Reply:
I think it's less brittle to not depend on the ordering of the values to match what was configured. If we need to change the interface to return an error, then I would return a new method for those drivers that want to use this customization

Reply:
I revised the API such that `RecordMetrics` is as before and additional labels get injects as in Prometheus with a wrapper. I also added tests.